components/browser/icons/src/main/java/mozilla/components/browser/icons/loader/HttpIconLoader.kt

```diff
@@ -42,13 +48,50 @@ class HttpIconLoader(
 
         val response = httpClient.fetch(downloadRequest)
 
-        if (!response.isSuccess) {
+        return if (response.isSuccess) {
+            response.toIconLoaderResult()
+        } else {
+            failureCache.rememberFailure(resource.url)
             return IconLoader.Result.NoResult
         }
+    }
 
-        return response.body.useStream {
-            IconLoader.Result.BytesResult(it.readBytes(),
-                Icon.Source.DOWNLOAD)
-        }
+    private fun shouldDownload(resource: IconRequest.Resource): Boolean {
+        return resource.url.toUri().isHttpOrHttps && !failureCache.hasFailedRecently(resource.url)
+    }
+}
+
+private fun Response.toIconLoaderResult() = body.useStream {
+    IconLoader.Result.BytesResult(it.readBytes(), Icon.Source.DOWNLOAD)
+}
+
+private const val MAX_FAILURE_URLS = 25
+private const val FAILURE_RETRY_MILLISECONDS = 1000 * 60 * 30 // 30 Minutes
+
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+internal class FailureCache {
+    private val cache = LruCache<String, Long>(MAX_FAILURE_URLS)
+
+    /**
+     * Remembers this [url] after loading from it has failed.
+     */
+    fun rememberFailure(url: String) {
+        cache.put(url, now())
     }
+
+    /**
+     * Has loading from this [url] failed previously and recently?
+     */
+    fun hasFailedRecently(url: String) =
+        cache.get(url)?.let { failedAt ->
```

Comment:
One thing I missed in my first pass is that `HttpIconLoader` runs on 3 worker threads. `LruCache` is thread-safe but we'd need to synchronize if we want to run multiple operations atomically. For the use case here I think that is fine though. The consequence is that we could remove a url that is no longer there (no effect) or remove one that failed more recently in case we have mutliple thread running for the same url.

Reply:
Good catch! I agree, in this case we may not need atomic operations.